Ignore unhandled Qt events

gselzer opened the pull request:
Currently, the Qt backend eats all events, regardless of whether vispy actually acts upon them. In my judgement, this seems contrary to the "best practices" of Qt, where events should be passed along if they are not acted upon.

This changeset calls [`QEvent.ignore()`](https://doc.qt.io/qt-6/qevent.html#ignore) in all of the relevant locations I could find (if there are some I missed, we should probably add this code there too), allowing parent widgets the chance to respond to unhandled events.

I'm wary about this change introducing other bugs, specifically in parent Qt widgets who are now receiving events they were not expecting - this changeset is likely a tradeoff that I think is worth consideration.

I'm not terribly sure whether this changeset warrants any changes to test code, and/or where those changes would be. If you'd like me to write tests, I'm happy to do so, I'd just appreciate a little direction on where to start!

brisvag commented:
Just did a quick test over at napari using this PR; I get a lot of:

```py
AttributeError: 'NoneType' object has no attribute 'handled'
```

I tried adding `event is not None` before very check and it seems enough. But I haven't tested more thoroughly than that.

gselzer commented:
> Just did a quick test over at napari using this PR; I get a lot of:
> 
> ```python
> AttributeError: 'NoneType' object has no attribute 'handled'
> ```
> 
> I tried adding `event is not None` before very check and it seems enough. But I haven't tested more thoroughly than that.

Hmm, yeah, that's unfortunate, thanks for checking @brisvag. We can definitely add the `None` checks here - my concern is more how the event becomes `None`, because it is never `None` when `BaseCanvasBackend._vispy_mouse_move` completes - seems like a race condition thing?

aganders3 commented:
`_vispy_mouse_move` will return `None` when it is within 10ms of the last event, I think:

https://github.com/gselzer/vispy/blob/cc6f5dbe45203276b7ef58859066ecaba0a48c85/vispy/app/base.py#L199-L200

gselzer commented:
> `_vispy_mouse_move` will return `None` when it is within 10ms of the last event, I think:
> 
> https://github.com/vispy/vispy/blob/dad688355f2c21c34cd04e4f639a326fe3c4ac1b/vispy/app/base.py#L199-L200

Oh man, how did I miss that - thanks @aganders3!

In that case, maybe we can just check for `None`s for mouse moves only? How then should a `None` be handled? I'd think it should also be `ignore()`d, in case the parent would like every Qt mouse moving event...

brisvag commented:
Seems reasonable to me!

gselzer commented:
> I'm a bit confused about all the different uses of "is None" vs "is not None" and different logic with "and" and "or"... Why are all these cases handled differently?

Yup, looks like some of the checks were handled differently...this should be solved with 98a5868, and I added in some clarifying comments. Thanks for bringing this up again @brisvag!

Most event handlers always return a vispy Event - in those cases we just need to ignore the qt event if `not event.handled` holds.

Some other handlers sometimes return `None`, which is equivalent to ignoring the event - in those cases we need more, hence the `event is None or not event.handled` logic. There were some (one?) `and` checks, which I also refactored to conform to this `or` logic.

brisvag commented:
@Czaki these chekcs being marked as fail though they were skipped - that's a mistake right? Should be good to merge?

Czaki commented:
You ask about cibuildwheel? 
I do not understand why ubuntu 20.04 is canceled.

brisvag commented:
Yeah. I also don't, I tried to rerun and it doesn't go :/

Czaki commented:
> Yeah. I also don't, I tried to rerun and it doesn't go :/ @djhoese, thoughts?

Ah, ubuntu-20.04. Created fix #2664.

djhoese commented:
Yep, from the summary page:

> This is a scheduled Ubuntu 20.04 brownout. Ubuntu 20.04 LTS runner will be removed on 2025-04-15. For more details, see https://github.com/actions/runner-images/issues/11101

brisvag commented:
We can merge this, the fails were unrelated and fixed in #2664!
